### PR TITLE
Kafka streaming unifier module

### PR DESCRIPTION
**Pull request by Ramizjon**

### CODE REVIEW

File: kafka-producer/src/main/scala/KafkaProducer.scala [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+    for (events <- Range(0, numEvents)) {
+      Thread.sleep(1000)
+      val message = outputGenerator.generateData()
+      val data = new KeyedMessage[String, String](topic, message)
+      producer.send(data)
+    }
+  }
+
+  def getGenerator(providerType: String): OutputGenerator = providerType match {
```
Comment (btrofimov) [inline kafka-producer/src/main/scala/KafkaProducer.scala]: public?

File: kafka-producer/src/main/scala/KafkaProducer.scala [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+import kafka.producer.{KeyedMessage, Producer, ProducerConfig}
+
+object KafkaProducer {
+
+  val numEvents = 50
+  val brokers = "localhost:9092"
+  val props = initProperties()
+
+
+  def initProperties(): Properties = {
```
Comment (btrofimov) [inline kafka-producer/src/main/scala/KafkaProducer.scala]: public?

File: kafka-producer/src/main/scala/utils/OutputGenerator.scala [hunk 3]
```diff
@@ -0,0 +1,21 @@
+package utils
+import scala.util.Random
+
+trait OutputGenerator {
+
+  def randomAlpha(length: Int): String = {
```
Comment (btrofimov) [inline kafka-producer/src/main/scala/utils/OutputGenerator.scala]: protected

File: kafka-producer/src/main/scala/utils/OutputGenerator.scala [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+    val chars = ('a' to 'z') ++ ('A' to 'Z')
+    randomStringFromCharList(length, chars)
+  }
+
+  def randomStringFromCharList(length: Int, chars: Seq[Char]): String = {
```
Comment (btrofimov) [inline kafka-producer/src/main/scala/utils/OutputGenerator.scala]: private

File: kafka-producer/src/main/scala/utils/OutputGenerator.scala [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+    val sb = new StringBuilder
+    for (i <- 1 to length) {
+      val randomNum = Random.nextInt(chars.length)
+      sb.append(chars(randomNum))
+    }
+    sb.toString
+  }
+
+  def generateData(): String
```
Comment (btrofimov) [inline kafka-producer/src/main/scala/utils/OutputGenerator.scala:21]: protected

File: streaming-unifier/src/main/scala/StreamingAppContext.scala [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+    def convertorMatcher(x: String): Convertor = x match {
+      case "nexus" => NexusConvertor
+      case "facebook" => FacebookConvertor
+    }
+  }
+
+  def runStreamingJobs(config: AppConfig): Unit = {
+    val sparkConf = new SparkConf()
+      .setAppName("directkafkaunifier")
+    val ssc = new StreamingContext(sparkConf, Duration.apply(10000))
```
Comment (btrofimov) [inline streaming-unifier/src/main/scala/StreamingAppContext.scala]: ugly, lets refactor it to something like `10 seconds`

**Comment by btrofimov** [comment]
please create `common` project

File: streaming-unifier/src/main/scala/driver/RDDProcessor.scala [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+import org.apache.spark.rdd.RDD
+import org.cloudera.spark.streaming.kafka.KafkaWriter._
+import org.slf4j.LoggerFactory
+import utils.{RDDTransformer, UserModCommand, Convertor, UMCKryoEncoder}
+
+trait LogTrait {
+  protected lazy val logger = LoggerFactory.getLogger(getClass)
+}
+
+abstract class RDDProcessor extends LogTrait with Serializable {
```
Comment (btrofimov) [inline streaming-unifier/src/main/scala/driver/RDDProcessor.scala]: with Serializable ?
Make RDDTransformer as a trait

**Comment by Ramizjon** [comment]
@btrofimov refactored some code accordingly last comment

File: common/src/main/scala/utils/RDDTransformer.scala [hunk 8]
```diff
@@ -0,0 +1,11 @@
+package utils
+
+import org.apache.spark.rdd.RDD
+
+trait RDDTransformer {
+  def transformRDD(input: RDD[String], conv: Convertor): RDD[UserModCommand] = {
+    input
+      .map(line => conv.convert(line))
```
Comment (btrofimov) [inline common/src/main/scala/utils/RDDTransformer.scala]:
```
.map(conv.convert(_))
```

File: common/src/main/scala/nexusprovider/NexusConvertor.scala [hunk 9]
```diff
@@ -1,22 +1,21 @@
...
 object NexusConvertor extends Convertor {
 
-  def convert (text: String): List[UserModCommand] = {
+  def convert(text: String): List[UserModCommand] = {
 
     val arr = text.split(",")
     val segmentsMap = arr
-      .view(3,arr.size)
+      .view(3, arr.size)
       .map { v =>
```
Comment (btrofimov) [inline common/src/main/scala/nexusprovider/NexusConvertor.scala]: _ -> arr(0)

**Comment by Ramizjon** [comment]
@btrofimov updated PR

**Comment by Ramizjon** [comment]
@btrofimov pull request updated

File: common/src/main/scala/nexusprovider/NexusConvertor.scala [hunk 10]
```diff
@@ -1,22 +1,21 @@
...
+  def convert(text: String): List[UserModCommand] = {
 
     val arr = text.split(",")
     val segmentsMap = arr
-      .view(3,arr.size)
-      .map { v =>
-      (v, arr(0))
+      .view(3, arr.size)
+      .map {
+        _ -> arr(0)
```
Comment (btrofimov) [inline common/src/main/scala/nexusprovider/NexusConvertor.scala:14]: Please avoid extended brackets if using undecore and simple expressions
Refactor to single line costruction

**Pull request merged**